operator/operator.go

```diff
@@ -115,150 +117,172 @@ func (operator *Operator) Serve(port int) error {
+	challenges := make(chan *bindings.ErdstallChallenged)
+	sub, err := operator.contract.WatchChallenged(nil, challenges, nil, nil)
+	if err != nil {
+		return fmt.Errorf("creating challenge subcription: %w", err)
+	}
+	defer sub.Unsubscribe()
 
-	return errg.Wait()
-}
+	for {
+		select {
+		case err := <-sub.Err():
+			return fmt.Errorf("subscription error: %w", err)
 
-type depositProofs struct {
-	mu      sync.RWMutex
-	entries map[common.Address]*tee.DepositProof
-}
+		case c := <-challenges:
+			log.Warnf("Operator.handleChallenges: Incoming challenge for account %v and epoch %v", c.Account.String(), c.Epoch)
 
-func newDepositProofs() *depositProofs {
-	return &depositProofs{entries: make(map[common.Address]*tee.DepositProof)}
-}
+			tr, err := func() (*bind.TransactOpts, error) {
+				ctx, cancel := createDefaultContext()
+				defer cancel()
 
-// Get gets the deposit proof for the given user, threadsafe.
-func (dps *depositProofs) Get(user common.Address) (*tee.DepositProof, bool) {
-	dps.mu.RLock()
-	defer dps.mu.RUnlock()
+				return operator.ethClient.NewTransactor(ctx, big.NewInt(0), eth.DefaultGasLimit, operator.ethClient.Account())
+			}()
+			if err != nil {
+				log.Errorf("Operator.handleChallenges: Failed to create transactor for response to challenge %v: %v", c, err)
+				continue
+			}
 
-	dp, ok := dps.entries[user]
+			balanceProof, ok := operator.balanceProofs.Get(c.Account)
+			if !ok {
+				log.Error("Operator.handleChallenges: Failed to get balance proof for challenge response")
+				continue
+			}
 
-	return dp, ok
-}
+			balance := bindings.ErdstallBalance{
+				Epoch:   balanceProof.Balance.Epoch,
+				Account: balanceProof.Balance.Account,
+				Value:   balanceProof.Balance.Value,
+			}
 
-// AddAll adds the given deposit proofs, threadsafe.
-func (dps *depositProofs) AddAll(in []*tee.DepositProof) {
-	dps.mu.Lock()
-	defer dps.mu.Unlock()
+			tx, err := operator.contract.Exit(tr, balance, balanceProof.Sig)
+			if err != nil {
+				log.Errorf("Operator.handleChallenges: Failed to send response to challenge %v: %v", c, err)
+				continue
+			}
 
-	for _, dp := range in {
-		dps.entries[dp.Balance.Account] = dp
+			// Track challenge response transaction status
+			//TODO: do not create a new go routine for every transaction, run a dedicated go routine that tracks all?
+			go func(c *bindings.ErdstallChallenged) {
+				ctx, cancel := createOnChainContext()
+				defer cancel()
+
+				r, err := bind.WaitMined(ctx, operator.ethClient.ContractBackend, tx)
+				if err != nil {
+					log.Errorf("Operator.handleChallenges: Failed to wait for mining of response to challenge %v: %v", c, err)
+					return
+				}
+
+				if r.Status != types.ReceiptStatusSuccessful {
+					//TODO: extract and log failure reason
+					log.Errorf("Operator.handleChallenges: Failed to complete response transaction for challenge %v", c)
+					return
+				}
+
+				log.Infof("Operator.handleChallenges: Resolved dispute for challenge %v", c)
```

Comment:
I just had a quick look at it after rebasing my client onto it, and it seems this does not format too well :laughing:.  
It outputs:  
```pre
INFO[0069] Operator.handleChallenges: Resolved dispute for challenge &{2 [18 166 197 36 199 32 117 157 39 82 54 165 114 81 52 235 28 71 73 221] {[79 184 99 122 253 40 73 42 50 9 1 117 86 233 93 194 248 8 109 219] [[159 113 104 110 158 46 237 10 10 153 52 11 28 59 35 3 105 242 85 177 212 82 19 12 234 213 79 131 8 101 77 253] [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2] [0 0 0 0 0 0 0 0 0 0 0 0 18 166 197 36 199 32 117 157 39 82 54 165 114 81 52 235 28 71 73 221]] [] 13 [195 37 53 68 40 35 185 152 139 79 218 227 201 66 145 88 129 56 47 74 202 233 131 209 9 207 60 114 112 198 47 111] 0 [82 70 186 155 7 187 173 137 50 186 127 193 185 233 248 18 56 45 170 36 190 75 227 75 247 197 80 132 83 249 54 44] 0 false}} 
```

Reply:
Added formatted printing.